src/async/Condition.php

```diff
@@ -66,9 +61,54 @@ final public function fail(\Exception $exception): void {
   final public async function waitForNotificationAsync(
     Awaitable<void> $notifiers,
   ): Awaitable<T> {
-    if ($this->condition === null) {
-      $this->condition = ConditionWaitHandle::create($notifiers);
-    }
-    return await $this->condition;
+    return await $this->state->waitForNotificationAsync($this, $notifiers);
   }
 }
+
+
+/**
+ * Asynchronously wait for the condition variable to be notified and
+ * return the result or throw the exception received via notification.
+ *
+ * The caller must provide an Awaitable $notifiers (which must be a
+ * WaitHandle) that must not finish before the notification is received.
+ * This means $notifiers must represent work that is guaranteed to
+ * eventually trigger the notification. As long as the notification is
+ * issued only once, asynchronous execution unrelated to $notifiers is
+ * allowed to trigger the notification.
+ */
+function wait_for_notification_async<T>(
+  (function(ConditionNotifyee<T>): Awaitable<void>) $notifiers,
+): Awaitable<T> {
+  $condition = new Condition();
+  return $condition->waitForNotificationAsync($notifiers($condition));
+}
```

Comment:
- at a minimum, this needs a similar 'do not use' comment to `Async\Poll`
- AIUI this is a control flow that we specificially do not want to make easier/more straightforward, due to ease of misuse (e.g. it can make it easy to make your request 10% faster by reducing server concurrent requests by 20%, in a way that doesn't show up in request-isolated benchmarking)